### PR TITLE
ddl: support add indexes for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

Problem Summary:

This PR supports add multiple indexes in one DDL statement.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -414,17 +414,30 @@ func (w *worker) updateDDLJob(t *meta.Meta, job *model.Job, meetErr bool) error
...
+	updateRawArgs := needToUpdateRawArgs(job, meetErr)
+	if !updateRawArgs {
 		logutil.Logger(w.logCtx).Info("[ddl] meet something wrong before update DDL job, shouldn't update raw args",
 			zap.String("job", job.String()))
-		updateRawArgs = false
 	}
 	return errors.Trace(t.UpdateDDLJob(0, job, updateRawArgs))
 }
 
+func needToUpdateRawArgs(job *model.Job, meetErr bool) bool {
```
Comment (xiongjiwei) [inline ddl/ddl_worker.go]: just use a  or `||` enough?

File: ddl/rollingback.go [hunk 2]
```diff
@@ -77,9 +77,9 @@ func convertAddIdxJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model.Job, tblIn
 	return ver, errors.Trace(err)
 }
 
-// convertNotStartAddIdxJob2RollbackJob converts the add index job that are not started workers to rollingbackJob,
+// convertNotReorgAddIdxJob2RollbackJob converts the add index job that are not started workers to rollingbackJob,
 // to rollback add index operations. job.SnapshotVer == 0 indicates the workers are not started.
-func convertNotStartAddIdxJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model.Job, occuredErr error) (ver int64, err error) {
+func convertNotReorgAddIdxJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model.Job, occuredErr error) (ver int64, err error) {
```
Comment (xiongjiwei) [inline ddl/rollingback.go:82]: why need change name, `NotStart` --> `NotReorg`?

Comment (tangenta) [reply ddl/rollingback.go:82]: `NotStart` means "the job is not started yet", which is inaccurate. It should be "the reorg worker is not running".

File: ddl/index.go [hunk 3]
```diff
@@ -469,12 +490,9 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 
 	if indexInfo == nil {
 		if len(hiddenCols) > 0 {
-			pos := &ast.ColumnPosition{Tp: ast.ColumnPositionNone}
-			for _, hiddenCol := range hiddenCols {
-				_, _, _, err = createColumnInfoWithPosCheck(tblInfo, hiddenCol, pos)
-				if err != nil {
-					job.State = model.JobStateCancelled
-					return ver, errors.Trace(err)
+			if len(hiddenCols) > 0 {
```
Comment (wjhuang2016) [inline ddl/index.go:493]: if len(hiddenCols) > 0 && if len(hiddenCols) > 0 ？

Comment (tangenta) [reply ddl/index.go]: The typo is fixed.

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/index.go [hunk 4]
```diff
@@ -599,6 +619,19 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
...
 
+func doReorgWorkForCreateIndexMultiSchema(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job,
+	tbl table.Table, indexInfo *model.IndexInfo) (done bool, ver int64, err error) {
+	if job.MultiSchemaInfo.Revertible {
+		done, ver, err = doReorgWorkForCreateIndex(w, d, t, job, tbl, indexInfo)
+		if done {
+			job.MarkNonRevertible()
+			done = false // We need another round to wait for all the others sub-jobs to finish.
+		}
+		return done, ver, err
```
Comment (zimulala) [inline ddl/index.go:630]: Could we do as following change?

```suggestion
		if done {
			job.MarkNonRevertible()
		}
		// We need another round to wait for all the others sub-jobs to finish.
		return false, ver, err
```

**Review by bb7133 — APPROVED** [review]
LGTM

File: ddl/index.go [hunk 5]
```diff
@@ -396,10 +396,31 @@ func checkPrimaryKeyNotNull(d *ddlCtx, w *worker, sqlMode mysql.SQLMode, t *meta
...
+	for i, c := range tblInfo.Columns {
+		if c.State != model.StatePublic {
+			firstNonPublicPos = i
+			break
+		}
+	}
+	for _, col := range idxInfo.Columns {
+		tblInfo.Columns[col.Offset].State = model.StatePublic
+		if _, needMove := hiddenColOffset[col.Offset]; needMove {
+			tblInfo.MoveColumnInfo(col.Offset, firstNonPublicPos)
```
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/index.go:423]: Got it. Move all hidden cols to the front of `firstNonPublicPos`.

File: ddl/multi_schema_change_test.go [hunk 6]
```diff
@@ -289,6 +289,84 @@ func TestMultiSchemaChangeAddDropColumns(t *testing.T) {
...
+	tk.MustGetErrCode("alter table t add index t(a, b), drop column a", errno.ErrUnsupportedDDLOperation)
+
+	// Test add index failed.
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int, b int, c int);")
+	tk.MustExec("insert into t values (1, 1, 1), (2, 2, 2), (3, 3, 1);")
+	tk.MustGetErrCode("alter table t add unique index i1(a), add unique index i2(a, b), add unique index i3(c);",
+		errno.ErrDupEntry)
+	tk.MustQuery("show index from t;").Check(testkit.Rows( /* no index */ ))
+	tk.MustExec("alter table t add index i1(a), add index i2(a, b), add index i3(c);")
```
Comment (zimulala) [inline ddl/multi_schema_change_test.go:325]: Do we need to add an expression index and add columns（test hidden cols or `moveAndUpdateHiddenColumnsToPublic`）in a statement?

File: ddl/multi_schema_change_test.go [hunk 7]
```diff
@@ -289,6 +289,84 @@ func TestMultiSchemaChangeAddDropColumns(t *testing.T) {
...
+	tk.MustExec("insert into t values (1, 2, 3);")
+	tk.MustExec("alter table t add index t(a, b), add index t1(a);")
+	tk.MustExec("alter table t add index t2(a), add index t3(a, b);")
+	tk.MustQuery("select * from t use index (t, t1, t2, t3);").Check(testkit.Rows("1 2 3"))
+	tk.MustExec("admin check table t;")
+
+	// Test add multiple indexes with same name.
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int, c int)")
+	tk.MustGetErrCode("alter table t add index t(a), add index t(b)", errno.ErrUnsupportedDDLOperation)
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/multi_schema_change_test.go:310]: I think we don't need to `admin check table`. The ErrUnsupportedDDL is reported by the DDL frontend. Moreover, there is no data and no index in this table.

File: ddl/multi_schema_change_test.go (hunk 6 above)
Comment (tangenta) [reply ddl/multi_schema_change_test.go:328]: Sure. I forgot to include the test for expression index in this PR.

File: ddl/multi_schema_change_test.go [hunk 8]
```diff
@@ -477,6 +480,46 @@ func TestMultiSchemaChangeAddDropIndexes(t *testing.T) {
...
+	tk.MustGetErrCode("alter table t add column c int, change column a d bigint, add index idx((a + a));", errno.ErrUnsupportedDDLOperation)
+	tk.MustGetErrCode("alter table t add column c int default 10, add index idx1((a + b)), add unique index idx2((a + b));",
+		errno.ErrDupEntry)
+	tk.MustQuery("select * from t;").Check(testkit.Rows("1 2", "2 1"))
+
+	originHook := dom.DDL().GetHook()
+	hook := &ddl.TestDDLCallback{Do: dom}
+	var checkErr error
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		assert.Equal(t, model.ActionMultiSchemaChange, job.Type)
```
Comment (zimulala) [inline ddl/multi_schema_change_test.go:500]: If `checkErr` isn't nil, we needn't do next logic, otherwise, the value of `checkErr` will be overwritten.

Comment (tangenta) [reply ddl/multi_schema_change_test.go:503]: Updated. Thanks!

**Pull request merged**